sycl/doc/extensions/experimental/sycl_ext_oneapi_virtual_mem.asciidoc

```diff
@@ -0,0 +1,343 @@
+access to features and gather community feedback.  Interfaces defined in this
+specification are implemented in {dpcpp}, but they are not finalized and may
+change incompatibly in future versions of {dpcpp} without prior notice.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Backend support status
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_oneapi_virtual_mem`.  The application must check that the devices
+in the corresponding context have this aspect before using any of the APIs
+introduced in this extension.  If the application fails to do this, the
+implementation throws a synchronous exception with the
+`errc::feature_not_supported` error code.
+
+== Overview
+
+This extension adds the notion of "virtual memory ranges" to SYCL, introducing
+a way to map an address range onto multiple allocations of physical memory,
+allowing users to avoid expensive reallocations and potentially running out of
+device memory while relocating the corresponding memory.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_VIRTUAL_MEM` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Memory granularity
+
+Working with virtual address ranges and the underlying physical memory requires
+the user to align and adjust in accordance with a specified minimum granularity.
+
+Some devices may have a recommended granularity that is different from the
+minimum granularity. In such cases, the reasoning for a different recommended
+granularity is unspecified. Devices may return the same minimum granularity as
+the recommended granularity.
+
+The interfaces for querying these granularities are defined as:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+size_t get_minimum_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext);
+size_t get_minimum_mem_granularity(size_t numBytes, const queue &syclQueue);
+size_t get_minimum_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem);
+
+size_t get_recommended_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext);
+size_t get_recommended_mem_granularity(size_t numBytes, const queue &syclQueue);
+size_t get_recommended_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem);
+
+} // namespace sycl::ext::oneapi::experimental
```

Comment:
This makes sense.  I think we should just clarify this for `get_recommended_mem_granularity` and say that the returned granularity could be less than, greater than, or equal to `numBytes`.

There's something about `get_minimum_mem_granularity` that seems weird.  For example, suppose the device supports page sizes of both 1024 and 4096.  If the user calls `get_minimum_mem_granularity(4096)`, what will they get?  Is the return value 1024 even though the requested `numBytes` doesn't fit in that granularity?  This makes me wonder if the minimum granularity should depend on `numBytes` at all.

Another weird thing is the word "minimum".  This word implies that the application could also choose a larger granularity, but that's not the case.  Presumably, the device supports a fixed set of granularities, and the application must choose one of them.  This makes me wonder if the API should instead just return a list of all the supported granularities like:

```
std::vector<size_t> get_mem_granularities(const device &syclDevice, const context &syclContext);
```

If the application wants the minimum one, they can just use the first element in the returned vector.

Reply:
Currently, the L0 interface corresponding to the granularity query ([zeVirtualMemQueryPageSize](https://spec.oneapi.io/level-zero/latest/core/api.html#zevirtualmemquerypagesize)) does not differentiate between recommended and minimum, but does take the size. Conversely, the CUDA query ([cuMemGetAllocationGranularity](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1g30ee906c2cf66a0347b3dfec3d7eb31a)) doesn't have the size argument, but has both a minimum and recommended mode.

Maybe we can remove the size from the minimum query by passing `1` to the L0 query and return the corresponding value, but it is not clear to me if that is actually always a valid granularity. @jandres742 - Do you know?

As for being able to return a list of valid granularities, I don't see how we can do that with the current L0 interfaces. For CUDA would have one or two elements (minimum and recommended, or one if they are the same.)

Reply:
right, @steffenlarsen . cuMemGetAllocationGranularity doesnt have a size parameter, so what that API does is:

"Here's the minimum and recommended granularities, please adjust your requested size to it"

While zeVirtualMemQueryPageSize says:

"For the size you want, here's the minimum granularity you should use for functionality and performance".

So the semantics of both APIs is different: CUDA's always returns the same numbers for a type of allocation, user needs to adjust the size, L0's already returns the granularity adjusted to the size. 

Now, the SYCL APIs proposed here are accepting a size, represented by numBytes

```cpp
sycl::get_recommended_mem_granularity(size_t numBytes...)
```

So I guess there's an expectation that the granularity returned by sycl::get_recommended_mem_granularity should take into account that size, which is what L0 is doing. So I dont think we should pass 1 to L0 API. What I think we should do is to modify cuda_piextVirtualMemGranularityGetInfo to not ignore the mem_size parameter, and instead, returned the granularity based on that size, something like:

```cpp
pi_result cuda_piextVirtualMemGranularityGetInfo(
    pi_context context, pi_device device, size_t mem_size,
    pi_virtual_mem_granularity_info param_name, size_t param_value_size,
    void *param_value, size_t *param_value_size_ret) {
...
      size_t granularity;
      result = PI_CHECK_ERROR(
          cuMemGetAllocationGranularity(&granularity, &alloc_props, flags));

       granularity = ROUND_UP(mem_size, granularity);

...
```

Now, if the intention of sycl::get_recommended_mem_granularity is to return a set of granularities, and have the user adjust the size, then current implementation in CUDA backend is ok, and for L0, a size of 1 could be passed.

So the main question here is : what is the intention of sycl::get_recommended_mem_granularity? is it to return the granularity based on the size passed (which is what L0 does, and for which we will need changes in the CUDA backend) or to ignore the size and return standard granularities (on which case it would be better to remove num_bytes from sycl::get_recommended_mem_granularity and to pass 1 to L0 API).

Reply:
One way to think about this is from the point of view of CUDA migration.  If a CUDA application calls `cuMemGetAllocationGranularity` with `CU_MEM_ALLOC_GRANULARITY_MINIMUM`, what do we expect the migrated code to call?  Same question for CUDA code that uses `CU_MEM_ALLOC_GRANULARITY_RECOMMENDED`.

Reply:
@gmlueck : in that case, for minimum and recommended, then yes,  a size of 1 could be passed to the L0 API. However, in that case, it would be better to remove the num_bytes parameter from sycl::get_recommended_mem_granularity. I mean, it is going to be ignored, then why it is needed?

Reply:
Thank you, @gmlueck & @jandres742 ! I had misunderstood the point of the memory size parameter of the page-size query in L0. For now it is best to keep the freedom of getting the granularity, so I have removed the memory size argument from the API. We can consider an interface for letting the backend align the address/sizes in the future.

Reply:
This latest change by @steffenlarsen aligns the SYCL API with CUDA, which makes migration easy, so that's good.

Does this cause us to lose some performance on Level Zero, though?  Let's say the user wants a moderately big (1Mb) address range.  With the current API, we'll call Level Zero `zeVirtualMemQueryPageSize` with `size` set to `1`.  Will this return a different answer than if we called it with `size` set to 1Mb?  If it is different, will this result is worse performance?

Reply:
An option is to add another PI function asking the backend to align for us. For CUDA we would just be using the recommended granularity and then L0 could work its magic. It means we would have somewhat similar APIs, but we get the best of both worlds as new users could just leverage this instead of doing their own aligning while people translating code have their 1:1 mapping in the existing functions.

Reply:
For PVC and ATS-M, zeVirtualMemQueryPageSize will return 64KBytes for any size less than 2MBytes, and will return 2MBytes for any size equal to or greater than 2MBytes.  Given this, entering size of 1 should probably not be an issue.